server/src/main/java/org/opensearch/cluster/ClusterModule.java

```diff
@@ -474,4 +475,9 @@ public void setExistingShardsAllocators(GatewayAllocator gatewayAllocator, Shard
         allocationService.setExistingShardsAllocators(existingShardsAllocators);
     }
 
+    public void setRerouteServiceForAllocator(RerouteService rerouteService) {
+        if (shardsAllocator instanceof BalancedShardsAllocator) {
+            ((BalancedShardsAllocator) shardsAllocator).setRerouteService(rerouteService);
+        }
+    }
```

Comment:
Can we have a default implementation in ShardAllocator and override the behaviour in BSA?

Reply:
Made changes. Can you check if this looks good?